Updating some RADIUSS packages from LLNL/radiuss-spack-configs

adrienbernede opened the pull request:
## Summary

We have been maintaining a "local" set of packages for a long time. We kept them up-to-date with changes in Spack at a slower pace, while they have improvements that will benefits others.

We are now in sync with Spack, so it’s a good time to submit this PR.

## Changes

- Some packages moved to the CachedCMakePackage class.
- Some packages have newer versions, new patches, conflicts, variants...
- There is a new package (raja-perf).
- Most packages are tuned to work well on LC system, relying on utility functions that are shared instead of duplicated, and ~located in the Camp package (1)~.

~## (1) Handling utility functions (located in Camp package)~

~We are of course open to suggestions regarding the best way to handle those utility functions.~

~Options appears to be:~
~1. Create a Radiuss (or LLNL) utility package.~
~2. Create a Radiuss (or LLNL) build system. ~
~3. Do nothing and submit upstream as-is, waiting for the appropriate capability in Spack.~
4. **Per @tgamblin suggestion: I am moving the utility functions from camp package to CachedCMake build system.**

-> Currently curating the utility functions. Solving conflicts with some logic already merged in https://github.com/spack/spack/pull/37592/files. Testing changes in https://github.com/LLNL/CHAI/pull/237.

adrienbernede committed:
Updating some RADIUSS packages with their counterpart from LLNL/radiuss-spack-configs

adrienbernede committed:
Add missing patch

adrienbernede committed:
Add raja-perf package

tldahlgren commented on var/spack/repos/builtin/packages/caliper/package.py:
```diff
@@ -101,63 +103,94 @@ class Caliper(CMakePackage, CudaPackage, ROCmPackage):
+        return sys_type
+
     def cmake_args(self):
+        options = []
+        return options
```
You could just `return []` unless you're expecting to add arguments here.

adrienbernede replied:
That, and the fact that cmake_args is defined twice in this package...

adrienbernede replied:
Fixed

tldahlgren commented on var/spack/repos/builtin/packages/raja-perf/package.py:
```diff
@@ -0,0 +1,243 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
```
```suggestion
# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```

tldahlgren commented on var/spack/repos/builtin/packages/raja/package.py:
```diff
@@ -1,12 +1,17 @@
-# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
```
```suggestion
# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```

tldahlgren commented on var/spack/repos/builtin/packages/umpire/package.py:
```diff
@@ -1,15 +1,19 @@
-# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
```
```suggestion
# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```

tldahlgren commented on var/spack/repos/builtin/packages/camp/package.py:
```diff
@@ -1,21 +1,32 @@
-# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
```
```suggestion
# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```

tldahlgren commented on var/spack/repos/builtin/packages/chai/package.py:
```diff
@@ -1,11 +1,17 @@
-# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
```
```suggestion
# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```

tldahlgren commented on var/spack/repos/builtin/packages/raja-perf/package.py:
```diff
@@ -0,0 +1,243 @@
+class RajaPerf(CachedCMakePackage, CudaPackage, ROCmPackage):
+    """RAJA Performance Suite."""
+
+    homepage = "http://software.llnl.gov/RAJAPerf/"
```
I cannot confirm this homepage.  Also, we prefer use of `https` protocol.

adrienbernede replied:
Agreed.

adrienbernede replied:
And fixed

tldahlgren reviewed (commented):
Confirmed all of the branches, tags, and sha256.  Suggestion for the homepage and copyright ending years should be `2023` (or at least not < `2022`).

adrienbernede committed:
Apply suggestions from code review: copyright year

Co-authored-by: Tamara Dahlgren <35777542+tldahlgren@users.noreply.github.com>

adrienbernede committed:
Apply review suggestions: homepages and shorter syntax

adrienbernede committed:
Fix : no more benchmarks variant in chai, use conflict syntax

adrienbernede committed:
Apply some of the formatting suggestions

tldahlgren commented on var/spack/repos/builtin/packages/chai/package.py:
```diff
@@ -106,51 +121,73 @@ def cache_name(self):
-            entries.insert(0, cmake_cache_path("CMAKE_CXX_COMPILER", spec["hip"].hipcc))
+
+        # adrienbernede-22-11:
+        #   This was in upstream Spack raja package, but it’s causing the follwing failure Umpire:
```
```suggestion
        #   This was in upstream Spack raja package, but it’s causing the following failure for Umpire:
```

tldahlgren commented on var/spack/repos/builtin/packages/raja-perf/package.py:
```diff
@@ -0,0 +1,248 @@
+    version("0.7.0", tag="v0.7.0", submodules="True")
+    version("0.6.0", tag="v0.6.0", submodules="True")
+    version("0.5.2", tag="v0.5.2", submodules="True")
+    version("0.5.1", tag="v0.5.1", submodules="True")
+    version("0.5.0", tag="v0.5.0", submodules="True")
+    version("0.4.0", tag="v0.4.0", submodules="True")
```
Should've mentioned this before but I didn't dig into the problem.  I get `All fetchers failed` for versions `0.4.0` and `0.5.1`.  

I briefly looked at the site last week and it seemed like they should work.

adrienbernede replied:
It turns out 0.4.0 and 0.5.1 miss the "v".

I’m fixing this.

<img width="279" alt="Screen Shot 2023-08-28 at 7 05 34 PM" src="https://github.com/spack/spack/assets/51493078/9a80af0c-bd91-425a-b800-6ed20796f0d8">

adrienbernede replied:
Fixed

tldahlgren commented:
> @tldahlgren Do we absolutely have to comply to the automated formatting ? Here is an example of the requested changes that are not for the best IMO:
> 
> <img alt="Screen Shot 2023-08-28 at 3 33 46 PM" width="645" src="https://user-images.githubusercontent.com/51493078/263718462-ed730fd5-9482-432b-94fb-c8839059b1b4.png">

They are required checks that I do not have the ability to override.  

Someone else complained recently about the value used for the line length but Todd G. pointed out the data he used to determine the value.

adrienbernede committed:
Update var/spack/repos/builtin/packages/chai/package.py

Co-authored-by: Tamara Dahlgren <35777542+tldahlgren@users.noreply.github.com>

adrienbernede committed:
Fix use existing tag

adrienbernede committed:
Simplify rocm support logic

adrienbernede committed:
Merge branch 'develop' into llnl/radiuss-packages-update

adrienbernede commented:
Per @tgamblin suggestion: I am moving the utility functions from camp package to CachedCMake build system.

Ongoing:
- Curating the utility functions.
- Solving conflicts with some logic already merged in https://github.com/spack/spack/pull/37592/files.
- Testing changes in https://github.com/LLNL/CHAI/pull/237.

adrienbernede committed:
Fix syntax in umpire package

adrienbernede committed:
Remove function that do not exists anymore

adrienbernede committed:
Apply suggestion to use a cmake string for a list of paths ()

https://github.com/spack/spack/pull/37592/files#r1213004352

adrienbernede committed:
Fix: remove unexisting function from import (only in umpire)

adrienbernede committed:
Complete removal of hip_for_radiuss_projects

adrienbernede committed:
Fix: add missing utility functions

adrienbernede committed:
Fix: add missing import

adrienbernede committed:
Move Cuda support utility functions to CachedCMakePackages

adrienbernede committed:
Fix: change variable name

adrienbernede committed:
Fix: change variable name

adrienbernede committed:
Remove CMAKE_PREFIX_PATH definition in cached_cmake package
This seems to be the only change remaining to explain failure to find rocm libs

adrienbernede committed:
Move BLT related utility functions to BLT package: too llnl specific to be part of the build system

adrienbernede commented:
@spack-bot commands

adrienbernede commented:
@spackbot commands

adrienbernede commented:
@spackbot fix style

adrienbernede commented:
@white238 I move a blt related function to the blt package instead of the cached cmake build-system.
Since it’s very LLNL and BLT specific, it made more sense to me. Thoughts?

adrienbernede commented:
@tldahlgren Is there a way to get the bot to fix style even in a PR from inside Spack/spack ?

alalazo commented:
@spackbot fix style

adrienbernede committed:
Apply isort and black formatting suggestions

adrienbernede committed:
Apply flake8 style corrections

adrienbernede committed:
Fix and complete black and flake8 reformat

adrienbernede committed:
Merge branch 'develop' into llnl/radiuss-packages-update

adrienbernede committed:
Fix missing "/" in paths

adrienbernede committed:
Add another gcc 4.9 path to the exclusion list

adrienbernede committed:
Merge branch 'develop' into llnl/radiuss-packages-update

adrienbernede committed:
removing local patch in favor of remote one with PR context

adrienbernede committed:
Adding an mpi variant to camp to export BLT MPI targets

This is a workaround to the issue where BLT targets are exported by both
Umpire (or RAJA) and Camp, where camp does not produce the MPI target,
which results in conflicts in projects importing from both.

adrienbernede committed:
Fix: overspeeding, no mpi variant in RAJA

adrienbernede committed:
TEMP: Ugly hack, for testing purpose only

adrienbernede committed:
Revert ugly hack

adrienbernede committed:
Fix style

adrienbernede committed:
Apply small patch to ascent

adrienbernede commented on var/spack/repos/builtin/packages/camp/package.py:
```diff
@@ -64,33 +53,72 @@ def cmake_args(self):
-        else:
-            options.append("-DENABLE_CUDA=OFF")
 
+        options.append(self.define_from_variant("ENABLE_HIP", "hip"))
```
```suggestion
        options.append(self.define_from_variant("ENABLE_HIP", "rocm"))
```

adrienbernede committed:
Fix variant name

adrienbernede commented on var/spack/repos/builtin/packages/camp/package.py:
```diff
@@ -64,33 +53,72 @@ def cmake_args(self):
+            else:
+                # starting with cmake 3.10, FindMPI expects MPIEXEC_EXECUTABLE
+                # vs the older versions which expect MPIEXEC
+                if self.pkg.spec["cmake"].satisfies("@3.10:"):
```
```suggestion
                if spec["cmake"].satisfies("@3.10:"):
```

adrienbernede committed:
Fix logic imported from super class

adrienbernede commented on var/spack/repos/builtin/packages/camp/package.py:
```diff
@@ -64,33 +53,72 @@ def cmake_args(self):
+                    mpiexec = os.path.join(spec["mpi"].prefix.bin, "mpiexec")
+
+            if not os.path.exists(mpiexec):
+                msg = "Unable to determine MPIEXEC, %s tests may fail" % self.pkg.name
```
```suggestion
                msg = "Unable to determine MPIEXEC, %s tests may fail" % self.name
```

adrienbernede committed:
Fix logic imported from super class

jhdavis8 commented:
I'm not sure what state this update is in but I get the following error when trying to build `raja+rocm~examples~exercises amdgpu_target=gfx90a generator=ninja %gcc@11.2.0` on OLCF Frontier using this branch:
```
==> Installing raja-2023.06.1-nqu7y2mnoh5qi5sksrbgdc4ucgqlpa6y [29/30]
==> No binary for raja-2023.06.1-nqu7y2mnoh5qi5sksrbgdc4ucgqlpa6y found: installing from source
==> Using cached archive: /autofs/nccs-svm1_home1/jhdavis/spack/var/spack/cache/_source-cache/git//LLNL/RAJA.git/9b5f61edf3aa1e6fdbc9a4b30828c81504639963.tar.gz
==> Warning: Fetching from mirror without a checksum!
  This package is normally checked out from a version control system, but it has been archived on a spack mirror.  This means we cannot know a checksum for the tarball in advance. Be sure that your connection to this mirror is secure!
==> No patches needed for raja
==> raja: Executing phase: 'initconfig'
==> Error: AttributeError: 'CompilerSpec' object has no attribute 'cxx'

The 'raja' package cannot find an attribute while trying to build from sources. This might be due to a change in Spack's package format to support multiple build-systems for a single package. You can fix this by updating the build recipe, and you can also report the issue as a bug. More information at https://spack.readthedocs.io/en/latest/packaging_guide.html#installation-procedure

/autofs/nccs-svm1_home1/jhdavis/spack/lib/spack/spack/build_systems/cached_cmake.py:293, in initconfig_hardware_entries:
        290            hip_link_flags = ""
        291            if "%gcc" in spec or spec_uses_toolchain(spec):
        292                if "%gcc" in spec:
  >>    293                    gcc_bin = os.path.dirname(self.pkg.spec.compiler.cxx)
        294                    gcc_prefix = os.path.join(gcc_bin, "..")
        295                else:
        296                    gcc_prefix = spec_uses_toolchain(spec)[0]
```

adrienbernede commented:
@jhdavis8 thank you for testing this branch and providing feedback.

There was indeed an error we never spotted in our CI due to not covering that case. Great that you found it !

adrienbernede committed:
Small fix in cached_cmake

adrienbernede committed:
Fixes MPI support and shared variant in Caliper

adrienbernede committed:
No more benchmark variant in CHAI

adrienbernede committed:
Minor change

adrienbernede committed:
Touch Umpire package for rebuild

adrienbernede committed:
Merge branch 'develop' into llnl/radiuss-packages-update

adrienbernede committed:
Remove ~benchmarks variant from CHAI spec

adrienbernede committed:
Apply change to Umpire from local package

adrienbernede committed:
Do not enforce c++14 in raja before 0.14

adrienbernede committed:
Enforce umpire 2022.03 in ascent specs

adrienbernede committed:
Fix what’s likely a typo and will solve CI issues

adrienbernede committed:
Revert "Apply small patch to ascent"

This reverts commit bb7ac8a134cafd23fa9984b68b7c39542959b7b8.

adrienbernede committed:
Try another fix for dray

adrienbernede committed:
Merge branch 'develop' into llnl/radiuss-packages-update

adrienbernede committed:
Merge branch 'develop' into llnl/radiuss-packages-update

adrienbernede committed:
Fix chai spec after update

adrienbernede committed:
for older umpire in ascent to workaround cmake issues waiting for newer BLT

adrienbernede committed:
fix umpire version to use with ascent

adrienbernede committed:
Merge branch 'develop' into llnl/radiuss-packages-update

adrienbernede committed:
Other ascent specs to change

adrienbernede committed:
Other ascent specs to change

adrienbernede committed:
Cray-mpich always uses slurm (or flux) as scheduler

adrienbernede committed:
Enforce llvm-amdgpu clang when activating rocm support

adrienbernede committed:
Merge tag 'develop-2023-10-29' into llnl/radiuss-packages-update

Snapshot release 2023-10-29

adrienbernede committed:
Merge branch 'develop' into llnl/radiuss-packages-update

adrienbernede committed:
Fix style

adrienbernede committed:
Update RADIUSS packages: MPI support in RAJA Perf

adrienbernede committed:
cache string to cache path

adrienbernede commented:
@spackbot rebuild everything

adrienbernede committed:
copy .cmake cache file to artifacts

adrienbernede committed:
Fix style

adrienbernede committed:
Fix typo

adrienbernede committed:
Revert "Fix style"

This reverts commit c2c41ce3be24fb1a7ad374c49f124776e81a82a2.

adrienbernede committed:
Fix style

adrienbernede committed:
Revert changes to ci.py (.cmake in artifacts)

adrienbernede committed:
Revert "Revert changes to ci.py (.cmake in artifacts)"

This reverts commit 2202d9475b9745759c977c6b0a16477aa5328334.

adrienbernede committed:
Fix path to .cmake cache file

adrienbernede committed:
Useless change to force CI

adrienbernede committed:
Merge tag 'develop-2023-11-05' into llnl/radiuss-packages-update

Snapshot release 2023-11-05

adrienbernede committed:
Merge branch 'develop' into llnl/radiuss-packages-update

adrienbernede committed:
Fix style

adrienbernede committed:
fix chai variant, no mpi when rocm in umpire raja and caliper

adrienbernede committed:
RAJA has no mpi variant

adrienbernede committed:
Use generic way to prevent +mpi +rocm for umpire and caliper in spack CI

adrienbernede committed:
Use old trick for old version of rocm

adrienbernede committed:
Fix syntax

adrienbernede committed:
Fix syntax

adrienbernede committed:
Fix formatting

adrienbernede committed:
Fix syntax

adrienbernede committed:
Import Version

adrienbernede committed:
Fix formatting

adrienbernede committed:
...

adrienbernede committed:
Merge commit '2ececcd03e' into llnl/radiuss-packages-update

adrienbernede committed:
Clean CI stacks: use package requirements

adrienbernede committed:
Restore CMAKE_PREFIX_PATH setup

adrienbernede committed:
New attempt

adrienbernede committed:
Fix and reorder dependencies

adrienbernede committed:
Fixing RAJA camp dependency, ordering, and specifying type=build

adrienbernede committed:
Adding commit for 2023.06.0 version of Camp hoping to get it built by Spack CI

adrienbernede committed:
Re-introduce former logic

adrienbernede committed:
Add rocm include dir symetric to binary dir

adrienbernede committed:
Remove recent addition that overrides potential fix

adrienbernede committed:
Fix style

adrienbernede committed:
Fix style (comment glob import after commenting glob use)

adrienbernede committed:
Fix style

adrienbernede committed:
TEST: remove new logic for testing

adrienbernede committed:
Fix line too long

adrienbernede committed:
Restore CMAKE_CXX_COMPILER = hipcc as top level definition

adrienbernede committed:
Add mpi variant to CHAI + cosmetic changes

adrienbernede committed:
Minor change

adrienbernede committed:
Apply patches that allow using umpire@2023.06 in ascent

adrienbernede committed:
Remove requirement of umpire@2022.03 for ascent in CI

adrienbernede committed:
Cosmetic change

adrienbernede committed:
Comment section with concurrent definition of hip compiler (amdclang v.s. hipcc)

adrienbernede committed:
Revert "Comment section with concurrent definition of hip compiler (amdclang v.s. hipcc)"

This reverts commit 16944c327a1a9df3d5463b500ca94a86f8007af6.

adrienbernede committed:
comment to revert to develop version

adrienbernede committed:
Fix style

adrienbernede commented:
@spackbot rebuild everything

adrienbernede commented:
This is deprecated. I’ll create a new PR when I push missing changes to the RADIUSS packages.
